File: hrm_app/lib/data/models/attendance_model.dart
class AttendanceRecord {
  final DateTime date;
  final DateTime? clockInTime;
  final DateTime? clockOutTime;
  final String status;
  final bool isLate;
  final bool isEarlyLeave;
  final List<dynamic> sessions;
  final Duration breakDuration;
  final Duration totalWorkDuration;
  final int latePenaltyMinutes;
  final int shortfallPenaltyMinutes;
  final int totalPenaltyMinutes;

  AttendanceRecord({
    required this.date,
    this.clockInTime,
    this.clockOutTime,
    required this.status,
    this.isLate = false,
    this.isEarlyLeave = false,
    this.sessions = const [],
    this.breakDuration = Duration.zero,
    this.totalWorkDuration = Duration.zero,
    this.latePenaltyMinutes = 0,
    this.shortfallPenaltyMinutes = 0,
    this.totalPenaltyMinutes = 0,
  });

  // Helper to get duration
  Duration get workingDuration {
    if (totalWorkDuration > Duration.zero) return totalWorkDuration;
    // Fallback
    if (clockInTime != null && clockOutTime != null) {
      return clockOutTime!.difference(clockInTime!);
    }
    return Duration.zero;
  }
}

File: hrm_backend/src/models/attendance.model.js
const mongoose = require("mongoose");

const attendanceSchema = new mongoose.Schema(
  {
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      required: true,
    },
    date: {
      type: Date,
      required: true,
    },
    clockInTime: {
      type: Date,
      required: true,
    },
    clockOutTime: {
      type: Date,
    },
    sessions: [
      {
        checkInTime: { type: Date, required: true },
        checkOutTime: { type: Date },
        duration: { type: Number, default: 0 }, // in ms
      },
    ],
    totalBreakDuration: {
      type: Number,
      default: 0,
    },
    breakStartTime: {
      // Keeping for backward compatibility or simple break tracking if needed, though sessions array is better
      type: Date,
    },
    breakEndTime: {
      type: Date,
    },
    status: {
      type: String,
      enum: ["Present", "Absent", "Half Day", "Leave", "On Duty"],
      default: "Present",
    },
    totalDuration: {
      type: Number, // In milliseconds
      default: 0,
    },
    // New Penalty Fields
    lateArrivalPenaltyMinutes: {
      type: Number,
      default: 0,
    },
    shortfallPenaltyMinutes: {
      type: Number,
      default: 0,
    },
    totalPenaltyMinutes: {
      type: Number,
      default: 0,
    },
    notes: {
      type: String,
    },
  },
  {
    timestamps: true,
  }
);

// Compound index to ensure one record per user per day (optional but recommended)
attendanceSchema.index({ userId: 1, date: 1 }, { unique: true });

module.exports = mongoose.model("Attendance", attendanceSchema);
